File: LinkedLists/remove_elements.py
from typing import Optional
from LinkedLists.list_node import ListNode


class RemoveElements:
    def remove_elements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        sentinel = ListNode(0, head)
        current = head
        prev = sentinel
        while current:
            if current.val == val:
                prev.next = current.next
            else:
                prev = current
            current = current.next

        return sentinel.next

# Time Complexity: O(N)
# Space Complexity: O(1)

File: LinkedLists/RemoveElements.java
package LinkedLists;

public class RemoveElements {
  public ListNode removeElements(ListNode head, int val) {
    ListNode sentinel = new ListNode(0, head);
    ListNode current = head;
    ListNode prev = sentinel;
    while (current != null) {
      if (current.val == val) {
        prev.next = current.next;
      } else {
        prev = current;
      }
      current = current.next;
    }

    return sentinel.next;
  }
}

// Time Complexity: O(n)
// Space Complexity: O(1)